cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,912 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../detail/knn_merge_parts.cuh"
+#include <raft/core/serialize.hpp>
+#include <raft/linalg/add.cuh>
+#include <raft/util/cuda_dev_essentials.cuh>
+
+#include "../../utils/nccl_helpers.cuh"
+#define NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/ann_mg.hpp>
+#undef NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/common.hpp>
+
+namespace cuvs::neighbors::mg {
+using namespace cuvs::neighbors;
+using namespace raft;
+
+void check_omp_threads(const int requirements);
+
+template <typename AnnIndexType, typename T, typename IdxT>
+template <typename Accessor>
+void ann_interface<AnnIndexType, T, IdxT>::build(
+  raft::resources const& handle,
+  const cuvs::neighbors::index_params* index_params,
+  raft::mdspan<const T, matrix_extent<int64_t>, row_major, Accessor> index_dataset)
+{
+  if constexpr (std::is_same<AnnIndexType, ivf_flat::index<T, IdxT>>::value) {
```

Comment:
We should probably centralize this logic somewhere so that's not only in the mg namespace. We will need to do this for the tiered index also (and we still shouldn't be doing this as methods, but rather free functions).

Reply:
Honestly, I'd prefer to put it all in `common.hpp` to make it brutally simple for users. cuVS' public APIs are not header-only and they are just declaring functions and classes who's symbols are largely defined in the shared object so there's little to no overhead for having large headers in `include/`.

Reply:
There's a circular dependency issue that makes things very messy. The iface struct depends on the definitions of `ivf_flat`, `ivf_pq` and `cagra`, which themselves depends on what is defined in `common.hpp`. I would have to either put include statements in the middle of the file or do so many forward declarations that it would render the code unreadable. Because of this, I believe that the best solution is to keep the iface in a separate file for now. Hope that's ok, else we can still discuss this more offline.

Reply:
I don’t think bad smells in code should be mitigated this way. I think it’s worth expending a little extra effort to simplify the experience and layout of the public API. All headers exposed in the public API are intended to be included directly by users, and since cuVS is only strictly compiled, we should only be exposing a minimal set of includes that the user absolutely needs. 

Also, forward declaration is pretty much par for the course in traditional c++ libraries. Is there a reason the interface of the struct can’t be declared in common.h and the implementation defined in a separate source file? We really should be doing that everywhere, but I’ve been more lenient on that thus far since it doesn’t really impact compile time all that much. The more symbols we compile into the cuvs binary; the less users need to compile into their apps (so overall compile times are reduced) .

sorry if this seems nitpicky but cuVS is now in a different category of software than both cuml and raft- we are heavily focused on the ease of integration and we can do that now since it’s no longer header only.